File: src/typegraph/python/typegraph/runtimes/graphql.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from dataclasses import dataclass
from typing import List, Optional

from typegraph import t
from typegraph.gen.runtimes import (
    BaseMaterializer,
    Effect,
    GraphqlRuntimeData,
    MaterializerGraphqlQuery,
)
from typegraph.runtimes.base import Materializer, Runtime
from typegraph.sdk import runtimes


@dataclass
class QueryMat(Materializer):
    path: List[str]


@dataclass
class MutationMat(Materializer):
    path: List[str]


class GraphQLRuntime(Runtime):
    endpoint: str

    def __init__(self, endpoint: str):
        runtime_id = runtimes.register_graphql_runtime(
            GraphqlRuntimeData(endpoint=endpoint)
        )

        super().__init__(runtime_id)
        self.endpoint = endpoint

    def query(
        self, inp: "t.struct", out: "t.typedef", *, path: Optional[List[str]] = None
    ):
        mat_id = runtimes.graphql_query(
            BaseMaterializer(runtime=self.id, effect="read"),
            MaterializerGraphqlQuery(path=path),
        )

        return t.func(inp, out, QueryMat(mat_id, effect="read", path=path))

    def mutation(
        self,
        inp: "t.struct",
        out: "t.typedef",
        effect: Effect,
        *,
        path: Optional[List[str]] = None,
    ):
        mat_id = runtimes.graphql_mutation(
            BaseMaterializer(runtime=self.id, effect=effect),
            MaterializerGraphqlQuery(path=path),
        )

        return t.func(inp, out, MutationMat(mat_id, effect=effect, path=path))

File: src/typegraph/rust/typegraph/src/runtimes/graphql.rs
use crate::{
    t::{self, TypeBuilder},
    wasm::{
        self,
        core::{RuntimeId, TypeId},
        runtimes::{BaseMaterializer, Effect, GraphqlRuntimeData, MaterializerGraphqlQuery},
    },
    Result,
};

#[derive(Debug)]
pub struct GraphqlRuntime {
    id: RuntimeId,
}

impl GraphqlRuntime {
    pub fn new(endpoint: &str) -> Result<Self> {
        let data = GraphqlRuntimeData {
            endpoint: endpoint.to_string(),
        };

        let id = wasm::with_runtimes(|r, s| r.call_register_graphql_runtime(s, &data))?;

        Ok(Self { id })
    }

    pub fn query<I, O>(
        &self,
        inp: I,
        out: O,
        path: Option<impl IntoIterator<Item = impl ToString>>,
    ) -> Result<TypeId>
    where
        I: TypeBuilder,
        O: TypeBuilder,
    {
        let base = BaseMaterializer {
            runtime: self.id,
            effect: Effect::Read,
        };

        let mat = MaterializerGraphqlQuery {
            path: path.map(|path| path.into_iter().map(|p| p.to_string()).collect()),
        };

        let mat = wasm::with_runtimes(|r, s| r.call_graphql_query(s, base, &mat))?;

        t::func(inp, out, mat)?.build()
    }

    pub fn mutation<I, O>(
        &self,
        inp: I,
        out: O,
        path: Option<impl IntoIterator<Item = impl ToString>>,
    ) -> Result<TypeId>
    where
        I: TypeBuilder,
        O: TypeBuilder,
    {
        let base = BaseMaterializer {
            runtime: self.id,
            effect: Effect::Read,
        };

        let mat = MaterializerGraphqlQuery {
            path: path.map(|path| path.into_iter().map(|p| p.to_string()).collect()),
        };

        let mat = wasm::with_runtimes(|r, s| r.call_graphql_mutation(s, base, &mat))?;

        t::func(inp, out, mat)?.build()
    }
}
